design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,312 @@
+      - the literal `rook-ceph-object-subuser`
+      - the name of the object store resource
+      - the full name of the subuser (including the `username:`-prefix).
+* `[2]` The full name of the subuser (including the `username:`-prefix).
+* `[3]` The generated swift access secret.
+
+### Risks and Mitigation
+
+As long as the Object Store CRD changes are well thought out the
+overall risk is minimal.
+
+* If Swift is enabled by accident this could lead to an increased and
+  unexpected attack surface, especially since the authentication and
+  authorization for Swift and S3 may differ depending on how Ceph RGW
+  is configured. This is mitigated by requiring explicit configuration
+  to enable Swift.
+
+* A misconfigured Keystone integration may allow users to gain access
+  to objects they should not be authorized to access (e.g. if the
+  `rgw keystone accepted roles` setting is too broad). This will be
+  mitigated by proper documentation to make the operator aware of the
+  risk.
+
+* Ceph RGW allows to disable TLS verification when querying Keystone,
+  we deliberately choose not to expose this config option to the user.
+
+## Drawbacks
+
+* As shown in [#4754](https://github.com/rook/rook/issues/4754)
+  keystone can be integrated via config override so
+  it is not strictly necessary to support configuring it via the
+  Object Store CRD. Adding it to the CRD complicates things with
+  minor gain.
+
+## Alternatives
+
+* There is no workable alternative to extending the CRD for Swift
+  support.
+
+* Keystone support could be configured via Rook config override as shown
+  in [#4754](https://github.com/rook/rook/issues/4754).
+
+## Open Questions
+
+### How to support multiple APIs in one Gateway
+
+Having a single Ceph RGW deployment serve both S3 and Swift is both
+possible and often desirable for interoperability reasons. The current
+structure of the Object Store CRD makes it difficult to represent the
+configuration for this cleanly.
+
+Possible Solutions:
+
+* Run a second object store resource. The question is then how to use
+  the same pools (since the pools are derived from the resource name
+  which must be unique)
+
+* Simply allow a list in the type argument (but keep allowing strings
+  for backwards compatibility). The drawback here is that the gateway
+  config will then be cluttered by the config options pertaining to
+  different API types.
+
+* Break compatibility with the existing CRD and split up the
+  `gateway:` configuration.
+
+* Keep only the common things in `gateway:` and add an own config
+  section for s3 and swift (those could be subsections of `gateway:`
+  or of new section, e.g. `apis:`). The `type:` option seems to be currently
+  ignored anyway. The config values are then merged from the
+  `gateway:`, the `s3:` and the `swift:` sections, with the values in
+  the specific sections taking precedence.
```

Comment:
Will we be able to test swift from the CI? Our integration tests setup a simple one-node K8s cluster to run the tests. It's not strictly a requirement to get swift working in the integration tests, but unit tests would then be our only way to catch issues.

Reply:
That's a good point. Ideally, we can test swift with a swift client the same way we do for S3. Failing that option, we could rely on a daily Rook+OpenStack integration test run.

Reply:
Testing swift in the CI with a swift client should be straightforward. Testing the keystone integration will be more challenging and require some kind of integration test.